mods/explorer-ctrln-newfile.wh.cpp

```diff
@@ -0,0 +1,435 @@
+    std::wstring path = join(dir, base);
+    if (!PathFileExistsW(path.c_str())) return path;
+    for (int i = 2; i < 10000; ++i) {
+        wchar_t buf[64];
+        swprintf(buf, ARRAYSIZE(buf), L"file (%d)", i);
+        path = join(dir, buf);
+        if (!PathFileExistsW(path.c_str())) return path;
+    }
+    return join(dir, L"file_new");
+}
+
+static bool CreateEmptyFile(const std::wstring& fullPath) {
+    HANDLE h = CreateFileW(fullPath.c_str(), GENERIC_WRITE, FILE_SHARE_READ, nullptr, CREATE_NEW,
+                           FILE_ATTRIBUTE_NORMAL, nullptr);
+    if (h == INVALID_HANDLE_VALUE) return false;
+    CloseHandle(h);
+    return true;
+}
+
+static bool BindParentAndChildPIDL(const std::wstring& fullPath, IShellFolder** ppFolder,
+                                   PCUITEMID_CHILD* ppChild) {
+    if (!ppFolder || !ppChild) return false;
+    *ppFolder = nullptr;
+    *ppChild = nullptr;
+
+    PIDLIST_ABSOLUTE pidl = nullptr;
+    SFGAOF sf = 0;
+    if (FAILED(SHParseDisplayName(fullPath.c_str(), nullptr, &pidl, 0, &sf)) || !pidl) return false;
+
+    IShellFolder* pFolder = nullptr;
+    PCUITEMID_CHILD pChild = nullptr;
+    HRESULT hr = SHBindToParent(pidl, IID_PPV_ARGS(&pFolder), &pChild);
+    CoTaskMemFree(pidl);
+    if (FAILED(hr)) return false;
+
+    *ppFolder = pFolder;  // keep ref
+    *ppChild = pChild;    // pointer owned by parent bind
+    return true;
+}
+
+static void SelectAndRenameInWindow(HWND hwndExplorer, const std::wstring& fullPath) {
+    IWebBrowser2* wb = nullptr;
+    if (!GetExplorerBrowserByHWND(hwndExplorer, &wb) || !wb) return;
+
+    wchar_t folderPath[MAX_PATH * 2] = {0};
+    wcscpy_s(folderPath, fullPath.c_str());
+    PathRemoveFileSpecW(folderPath);
+
+    IServiceProvider* sp = nullptr;
+    if (FAILED(wb->QueryInterface(IID_PPV_ARGS(&sp))) || !sp) {
+        wb->Release();
+        return;
+    }
+
+    IShellBrowser* psb = nullptr;
+    HRESULT hr = sp->QueryService(SID_STopLevelBrowser, IID_PPV_ARGS(&psb));
+    sp->Release();
+    if (FAILED(hr) || !psb) {
+        wb->Release();
+        return;
+    }
+
+    IShellView* psv = nullptr;
+    hr = psb->QueryActiveShellView(&psv);
+    psb->Release();
+    if (FAILED(hr) || !psv) {
+        wb->Release();
+        return;
+    }
+
+    IShellFolder* pParent = nullptr;
+    PCUITEMID_CHILD pChild = nullptr;
+    if (BindParentAndChildPIDL(fullPath, &pParent, &pChild)) {
+        psv->SelectItem(pChild, SVSI_SELECT | SVSI_ENSUREVISIBLE | SVSI_DESELECTOTHERS | SVSI_EDIT);
+        pParent->Release();
+    }
+
+    psv->Release();
+    wb->Release();
+}
+
+// ----------------- New file action -----------------
+static void PerformNewFileAction() {
+    ComInit com;
+
+    HWND target = GetTargetExplorerHWND();
+    if (!target) {
+        Wh_Log(L"[Ctrln] No explorer window detected.");
+        return;
+    }
+
+    std::wstring dir;
+    if (!GetExplorerPathFromHWND(target, dir) || dir.empty()) dir = GetDesktopDir();
```

Comment:
If a path is unsupported (e.g. Control Panel, I assume), it seems wrong to randomly create a file on the desktop.